nab/optimizer.py

```diff
@@ -18,134 +18,303 @@
+  :returns: a tuple of (scoreChange, countsChange)
+  """
+  scoreChange = 0.0
+  countsChange = collections.defaultdict(int)
+  if dataInfo.lastWindow is None:
+    # False positive before any windows
+    if not dataInfo.probation:
+      # Probationary period has passed
+      scoreChange -= costMatrix["fpWeight"]
+
+      # Update countsChange
+      countsChange["tn"] -= 1
+      countsChange["fp"] += 1
+    else:
+      # Do nothing, don't count the FP since it is inside probation period
+      pass
+  elif dataInfo.idx > dataInfo.lastWindow.end:
+    # Outside window, contribute false positive
+    scoreChange += (computeScaledScoreOutsideWindow(dataInfo) *
+                    costMatrix["fpWeight"])
+
+    # Update countsChange
+    countsChange["tn"] -= 1
+    countsChange["fp"] += 1
+  elif len(dataInfo.lastWindow.detectedAnomalies) == 0:
+    # First anomaly detected inside the window
+    scoreChange += costMatrix["fnWeight"]
+    scoreChange += (computeScaledScoreInsideWindow(dataInfo) *
+                    costMatrix["tpWeight"])
+    # Update the last windows list of detected anomalies
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+  elif dataInfo.idx < dataInfo.lastWindow.detectedAnomalies[0]:
+    # Not first instance of a detected anomaly but earliest in window
+    prevDataInfo = dataInfo.lastWindow.detectedAnomalies[0]
+    scoreChange -= (computeScaledScoreInsideWindow(prevDataInfo) *
+                    costMatrix["tpWeight"])
+
+    scoreChange += (computeScaledScoreInsideWindow(dataInfo) *
+                    costMatrix["tpWeight"])
+
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+  else:
+    # Not first or earliest anomaly in window
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+
+  return scoreChange, countsChange
+
+
+
+def _computeThresholdScores(data, numWindows, costMatrix):
+  """Compute the scores for each threshold.
+
+  This is the core of the optimization. It takes the preprocessed data, sorts
+  it by anomaly score (descending), and iteratively lowers the threshold to
+  detect one more row as an anomaly. The threshold is originally set greater
+  than 1.0 and the score is set to the null detector score for the data. Then
+  the threshold is lowered to the anomaly score of each sorted row in turn.
+  Since some thresholds result in multiple rows being detected at once, we only
+  keep the score that corresponds to all of the newly-detected rows being
+  detected.
```

Comment:
What does this last sentence mean? Are you referring to the (common) case where multiple rows have the same anomaly score?